// @browserbasehq/sdk/_shims/auto/runtime-node.mjs
import { Readable } from 'node:stream';
import { getRuntime as getNodeRuntime } from "../node-runtime.mjs";
function usesNodeFetchOnlyFeatures(init) {
    if (!init)
        return false;
    // Node's native fetch does not use node:http Agent instances. Preserve
    // historical behavior for callers who explicitly configure httpAgent.
    return Boolean(init.agent);
}
function buildNativeFetchOptions(init) {
    if (!init)
        return init;
    const { agent: _agent, ...fetchInit } = init;
    // Node's native fetch requires `duplex: 'half'` when the body is a stream.
    const body = fetchInit.body;
    if (body instanceof Readable || typeof body?.pipe === 'function') {
        return { duplex: 'half', ...fetchInit };
    }
    return fetchInit;
}
export function getRuntime() {
    const nodeRuntime = getNodeRuntime();
    const nativeFetch = globalThis.fetch;
    if (typeof nativeFetch !== 'function') {
        return nodeRuntime;
    }
    return {
        ...nodeRuntime,
        fetch: (url, init) => {
            if (usesNodeFetchOnlyFeatures(init)) {
                return nodeRuntime.fetch(url, init);
            }
            return nativeFetch.call(undefined, url, buildNativeFetchOptions(init));
        },
        Request: typeof globalThis.Request !== 'undefined' ? globalThis.Request : nodeRuntime.Request,
        Response: typeof globalThis.Response !== 'undefined' ?
            globalThis.Response
            : nodeRuntime.Response,
        Headers: typeof globalThis.Headers !== 'undefined' ? globalThis.Headers : nodeRuntime.Headers,
        getDefaultAgent: () => undefined,
    };
}
