// @browserbasehq/sdk/_shims/auto/runtime-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRuntime = getRuntime;
/**
 * Disclaimer: modules in _shims aren't intended to be imported by SDK users.
 */
const node_stream_1 = require("node:stream");
const node_runtime_1 = require("../node-runtime.js");
function usesNodeFetchOnlyFeatures(init) {
    if (!init)
        return false;
    // Node's native fetch does not use node:http Agent instances. Preserve
    // historical behavior for callers who explicitly configure httpAgent.
    return Boolean(init.agent);
}
function buildNativeFetchOptions(init) {
    if (!init)
        return init;
    const { agent: _agent, ...fetchInit } = init;
    // Node's native fetch requires `duplex: 'half'` when the body is a stream.
    const body = fetchInit.body;
    if (body instanceof node_stream_1.Readable || typeof body?.pipe === 'function') {
        return { duplex: 'half', ...fetchInit };
    }
    return fetchInit;
}
function getRuntime() {
    const nodeRuntime = (0, node_runtime_1.getRuntime)();
    const nativeFetch = globalThis.fetch;
    if (typeof nativeFetch !== 'function') {
        return nodeRuntime;
    }
    return {
        ...nodeRuntime,
        fetch: (url, init) => {
            if (usesNodeFetchOnlyFeatures(init)) {
                return nodeRuntime.fetch(url, init);
            }
            return nativeFetch.call(undefined, url, buildNativeFetchOptions(init));
        },
        Request: typeof globalThis.Request !== 'undefined' ? globalThis.Request : nodeRuntime.Request,
        Response: typeof globalThis.Response !== 'undefined' ?
            globalThis.Response
            : nodeRuntime.Response,
        Headers: typeof globalThis.Headers !== 'undefined' ? globalThis.Headers : nodeRuntime.Headers,
        getDefaultAgent: () => undefined,
    };
}
